### ISSUE TITLE
How to convet qpos to actions?

### ISSUE THREAD

**Issue by Xianqi-Zhang** [carlosferrazza/humanoid-bench#21]
Hi,
Thanks for your sharing.
I am very sorry to bother you so many times, and I have been following this work for a long time.

My work requires to explicitly estimate the residual of fine-tuning from the current pose, so I need to convert qpos into action. Your previous reply said that the robot in then env is **position controlled**, so I tried to select the body-related variables in qpos as actions, but it doesn't seem right.

In order to achieve this function (qpos -> action), I first execute some actions, record data, and then execute actions generated by qpos, as follows,
```python
    # * Dims related to body parts in self._env.named.data.qpos.
    action = torch.zeros((61))
    action[:16] = ob['proprio'][:16]
    action[40:45] = ob['proprio'][40:45]
```

I also tried adding normalize_action()/unnormalize_action(), but it seems not right, because the two simulation results are very different.
```python
    action = torch.zeros((61))
    action[:16] = ob['proprio'][:16]
    action[40:45] = ob['proprio'][40:45]
    # * action = env.task.task.unnormalize_action(action)
    action = env.task.task.normalize_action(action)
```
Test settings：
```python
    parser.add_argument("--policy_type", default=None)
    parser.add_argument('--obs_wrapper', default='True')
    parser.add_argument('--sensors', default='image')
```
So the test env use ObservationWrapper, and the action is unnormalized before executed.
https://github.com/carlosferrazza/humanoid-bench/blob/376aec79a85ea5df855d1e0e72b00c750baafd7b/humanoid_bench/wrappers.py#L564-L565
https://github.com/carlosferrazza/humanoid-bench/blob/376aec79a85ea5df855d1e0e72b00c750baafd7b/humanoid_bench/tasks.py#L59-L61
**I think the qpos-action need to be normalized (as the oppsite of unnormalize), but the visualized results of unormalize_action seems better than normalize_action. This confuses me.**

A frame of my test seq. (Frame 05)
Src action.
<img src=https://github.com/user-attachments/assets/857805e4-96cc-41fd-8bc1-437de827c41d width=200 height=200 /> 

Directly-qpos
<img src=https://github.com/user-attachments/assets/a7d542fe-9f23-4966-8745-c5e778353501 width=200 height=200 /> 

Normalized-qpos
<img src=https://github.com/user-attachments/assets/b8fa40b2-b6d3-40ef-9ea1-59bd468ede06 width=200 height=200 /> 

Unnormalized-qpos
<img src=https://github.com/user-attachments/assets/f03b438d-c91b-4996-8444-a5ccac45d15a width=200 height=200 /> 

**I have no idea about:**
1. How to convert qpos to actions? Or which one is more reasonable, (directly-qpos, normalized-qpos, or unnormalize-qpos)? Or any other advices?
2. Is the difference between executing directly_qpos and action due to gravity and some random nature of the simulator? 
(It seems that: when an action is executed, the simulation is not until the robot and objects are stationary, it only simluate a certain steps. When performing the same action repeatedly, it usually visualizes a gradual fall.)
 

Thanks for your any reply.

Best regards.

**Comment by carlosferrazza** [comment]
Hi! I think you are getting confused by the order of joints and actuators in the xml. Try the following code in the `test_env.py` file:

    action_raw = np.zeros(env.action_space.shape)
    joint_to_action = {}
    offset = 0
    for i in range(env.model.njnt):
        print(f"joint {i}: {env.model.joint(i).name}")
        jnt_name = env.model.joint(i).name
        if jnt_name.startswith("free"):
            joint_to_action[env.model.joint(i).name] = range(i + offset, i + offset + 7)
            offset += 6
        else:
            joint_to_action[env.model.joint(i).name] = i + offset

    for i in range(env.model.nu):
        print(f"actuator {i}: {env.model.actuator(i).name}")
        act_name = env.model.actuator(i).name
        if act_name.startswith("lh") or act_name.startswith("rh"):
            action_raw[i] = 0.0
        else:
            action_raw[i] = env.data.qpos[joint_to_action[act_name]]

Then, you just need to normalize:

    action = env.task.normalize_action(action_raw)

Note that I am setting the hand joints to zero above.

**Comment by carlosferrazza** [comment]
The code I posted above is the correct way to control the body qpos at their current positions. The initial keyframe is not passively stable, that is why the robot eventually falls down.
